From e9102f755ffebc70c3c7277784d10f2d8090064e Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Wed, 19 Oct 2022 09:16:34 +0200
Subject: [PATCH 0754/1013] evl/poll: lockdep: force registration of dynamic
 lock class

Lockdep may register a lock class lazily at first lock acquisition,
which might happen oob for us. Since that registration depends on RCU,
we need to make sure this always happens in-band instead.

Use might_lock() to force lockdep to pre-register the lock class at
lock init time.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/poll.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/evl/poll.c b/kernel/evl/poll.c
index f0b7d63634a6..08bba4f4efba 100644
--- a/kernel/evl/poll.c
+++ b/kernel/evl/poll.c
@@ -605,6 +605,7 @@ int wait_events(struct file *filp,
 static int poll_open(struct inode *inode, struct file *filp)
 {
 	struct poll_group *group;
+	unsigned long flags;
 	int ret;
 
 	group = kzalloc(sizeof(*group), GFP_KERNEL);
@@ -622,6 +623,9 @@ static int poll_open(struct inode *inode, struct file *filp)
 	INIT_LIST_HEAD(&group->waiter_list);
 	evl_init_kmutex(&group->item_lock);
 	raw_spin_lock_init(&group->wait_lock);
+	local_irq_save(flags);
+	might_lock(&group->wait_lock); /* see __evl_init_wait(). */
+	local_irq_restore(flags);
 	filp->private_data = group;
 	stream_open(inode, filp);
 
-- 
2.30.2

